### PR TITLE
multiple subaccounts in one namespace

### PR DESCRIPTION
**Pull request by TalShorSap**
1. multiple subaccounts in one namespace
2. add flag to cluster config map: enableMultipleSubaccounts

### CODE REVIEW

File: internal/secrets/resolver_test.go [hunk 1]
```diff
@@ -37,6 +37,7 @@ var _ = Describe("Secrets Resolver", func() {
 		} else {
 			name = fmt.Sprintf("%s-%s", namePrefix, secrets.SAPBTPOperatorSecretName)
 		}
+		fmt.Println(fmt.Sprintf("Creating secret with name %s", name))
```
Comment (kerenlahav) [inline internal/secrets/resolver_test.go:40]: use log

Comment (kerenlahav) [inline internal/secrets/resolver_test.go:40]: use by

File: sapbtp-operator-charts/templates/crd.yml [hunk 2]
```diff
@@ -842,6 +848,9 @@ spec:
               externalName:
                 description: The name of the instance in Service Manager
                 type: string
+              subaccountID:
+                description: The subaccount id of the service instance
```
Comment (kerenlahav) [inline sapbtp-operator-charts/templates/crd.yml:852]: remove, not supported for v1alpha1

File: internal/secrets/resolver_test.go [hunk 3]
```diff
@@ -67,14 +68,21 @@ var _ = Describe("Secrets Resolver", func() {
...
 
 	validateSecretResolved := func() {
-		resolvedSecret, err := resolver.GetSecretForResource(ctx, testNamespace, secrets.SAPBTPOperatorSecretName)
+		resolvedSecret, err := resolver.GetSecretForResource(ctx, testNamespace, secrets.SAPBTPOperatorSecretName, "")
+		Expect(err).ToNot(HaveOccurred())
+		Expect(resolvedSecret).ToNot(BeNil())
+		Expect(string(resolvedSecret.Data["clientid"])).To(Equal(expectedClientID))
+	}
+
+	validateSubaccountSecretResolved := func(subaccountID string) {
```
Comment (kerenlahav) [inline internal/secrets/resolver_test.go:77]: if it is used only once and for a specific test the method is not needed, put the code in the test it is more readable

File: sapbtp-operator-charts/templates/crd.yml (hunk 2 above)
Comment (TalShorSap) [reply sapbtp-operator-charts/templates/crd.yml:852]: right

File: internal/secrets/resolver_test.go (hunk 3 above)
Comment (TalShorSap) [reply internal/secrets/resolver_test.go:77]: i thought it can be reused in the future

Comment (TalShorSap) [reply internal/secrets/resolver_test.go]: but i applied your comment

File: controllers/serviceinstance_controller.go [hunk 4]
```diff
@@ -309,6 +310,10 @@ func (r *ServiceInstanceReconciler) createInstance(ctx context.Context, smClient
 	log.Info("Instance provisioned successfully")
 	serviceInstance.Status.InstanceID = provision.InstanceID
 
+	if serviceInstance.Spec.SubaccountID != "" {
```
Comment (kerenlahav) [inline controllers/serviceinstance_controller.go:313]: there is no value in this, need to fetch the subaccount id from the context

File: api/v1/serviceinstance_validating_webhook.go [hunk 5]
```diff
@@ -34,19 +35,28 @@ func (si *ServiceInstance) SetupWebhookWithManager(mgr ctrl.Manager) error {
...
 
-// +kubebuilder:webhook:verbs=delete,path=/validate-services-cloud-sap-com-v1-serviceinstance,mutating=false,failurePolicy=fail,groups=services.cloud.sap.com,resources=serviceinstances,versions=v1,name=vserviceinstance.kb.io,sideEffects=None,admissionReviewVersions=v1beta1;v1
+// +kubebuilder:webhook:verbs=delete;update,path=/validate-services-cloud-sap-com-v1-serviceinstance,mutating=false,failurePolicy=fail,groups=services.cloud.sap.com,resources=serviceinstances,versions=v1,name=vserviceinstance.kb.io,sideEffects=None,admissionReviewVersions=v1beta1;v1
 
 var _ webhook.Validator = &ServiceInstance{}
 
+// log is for logging in this package.
+var serviceinstancelog = logf.Log.WithName("serviceinstance-resource")
+
 func (si *ServiceInstance) ValidateCreate() (warnings admission.Warnings, err error) {
```
Comment (kerenlahav) [inline api/v1/serviceinstance_validating_webhook.go:45]: don't allow to add subaccount id if enable_multiple_subaccounts is false

**Review by kerenlahav — APPROVED** [review]

File: api/v1/serviceinstance_types.go [hunk 6]
```diff
@@ -82,6 +82,9 @@ type ServiceInstanceSpec struct {
 	// end-user. User-provided values for this field are not saved.
 	// +optional
 	UserInfo *v1.UserInfo `json:"userInfo,omitempty"`
+
+	// The subaccount id of the service instance
```
Comment (evyaffe) [inline api/v1/serviceinstance_types.go:86]: The desired subaccount id for the service instance

File: api/v1/serviceinstance_validating_webhook.go [hunk 7]
```diff
@@ -34,19 +35,37 @@ func (si *ServiceInstance) SetupWebhookWithManager(mgr ctrl.Manager) error {
...
+var serviceinstancelog = logf.Log.WithName("serviceinstance-resource")
+var allowMultipleTenants bool
+
+func SetAllowMultipleTenants(isAllowed bool) {
+	allowMultipleTenants = isAllowed
+}
+
 func (si *ServiceInstance) ValidateCreate() (warnings admission.Warnings, err error) {
+	serviceinstancelog.Info("validate create", "name", si.Name)
+	if !allowMultipleTenants && si.Spec.SubaccountID != "" {
```
Comment (evyaffe) [inline api/v1/serviceinstance_validating_webhook.go:52]: len(si.Spec.SubaccountID) > 0

File: api/v1/serviceinstance_validating_webhook.go [hunk 8]
```diff
@@ -34,19 +35,37 @@ func (si *ServiceInstance) SetupWebhookWithManager(mgr ctrl.Manager) error {
...
+		return nil, fmt.Errorf("using multiple subaccounts is not allowed")
```
Comment (evyaffe) [inline api/v1/serviceinstance_validating_webhook.go:53]: log.Error(fmt.Errorf("invalid subaccountID property"), "the operator installation does not allow multiple subaccunts")
return nil, fmt.Errorf("setting the subaccountID property is not allowed")

File: api/v1/serviceinstance_validating_webhook.go [hunk 9]
```diff
@@ -34,19 +35,37 @@ func (si *ServiceInstance) SetupWebhookWithManager(mgr ctrl.Manager) error {
...
+	}
 	return nil, nil
 }
 
 func (si *ServiceInstance) ValidateUpdate(old runtime.Object) (warnings admission.Warnings, err error) {
+	serviceinstancelog.Info("validate update", "name", si.Name)
+	oldInstance := old.(*ServiceInstance)
+	if oldInstance.Spec.SubaccountID != si.Spec.SubaccountID {
+		return nil, fmt.Errorf("subaccountID can not be changed")
```
Comment (evyaffe) [inline api/v1/serviceinstance_validating_webhook.go:62]: return nil, fmt.Errorf("changing the subaccountID for an existing instance is not allowed")

File: controllers/servicebinding_controller.go [hunk 10]
```diff
@@ -89,7 +89,25 @@ func (r *ServiceBindingReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 		}
 	}
 
-	smClient, err := r.getSMClient(ctx, serviceBinding)
+	log.Info("service instance name " + serviceBinding.Spec.ServiceInstanceName + " binding namespace " + serviceBinding.Namespace)
```
Comment (evyaffe) [inline controllers/servicebinding_controller.go:92]: improve log

File: sapbtp-operator-charts/templates/configmap.yml [hunk 11]
```diff
@@ -22,6 +22,11 @@ data:
   ALLOWED_NAMESPACES: {{ join "," .Values.manager.allowed_namespaces }}
   {{- end }}
   {{- end }}
+  ENABLE_MULTIPLE_SUBACCOUNTS: {{ .Values.manager.enable_multiple_subaccounts | quote }}
+  {{- if not .Values.manager.enable_multiple_subaccounts }}
```
Comment (evyaffe) [inline sapbtp-operator-charts/templates/configmap.yml:26]: what is the meaning of the 2 "if" ?

File: internal/secrets/resolver.go [hunk 12]
```diff
@@ -20,68 +19,70 @@ const (
...
 	}
 
-	if !found {
-		// secret not found in resource namespace, search for namespace-specific secret in management namespace
-		sr.Log.Info("Searching for namespace secret in management namespace", "namespace", namespace, "managementNamespace", sr.ManagementNamespace, "name", name)
-		secretForResource, err = sr.getSecretForNamespace(ctx, namespace, name)
-		if client.IgnoreNotFound(err) != nil {
-			sr.Log.Error(err, "Could not fetch secret in management namespace")
+	// search subaccount secret
+	if subaccountID != "" {
```
Comment (evyaffe) [inline internal/secrets/resolver.go:38]: len(subaccountID) > 0

File: internal/secrets/resolver.go [hunk 13]
```diff
@@ -20,68 +19,70 @@ const (
...
-		sr.Log.Info("Searching for secret in resource namespace", "namespace", namespace, "name", name)
-		secretForResource, err = sr.getSecretFromNamespace(ctx, namespace, name)
-		if client.IgnoreNotFound(err) != nil {
-			sr.Log.Error(err, "Could not fetch secret in resource namespace")
-			return nil, err
-		}
-
-		found = !apierrors.IsNotFound(err)
+	if !sr.EnableMultipleSubaccounts {
+		return sr.getDefaultSecret(ctx, name)
```
Comment (evyaffe) [inline internal/secrets/resolver.go:34]: add log.info that EnableMultipleSubaccounts is disabled and therefore using default

File: controllers/servicebinding_controller.go (hunk 10 above)
Comment (TalShorSap) [reply controllers/servicebinding_controller.go:92]: that was for debugging

File: api/v1/serviceinstance_validating_webhook.go (hunk 9 above)
Comment (TalShorSap) [reply api/v1/serviceinstance_validating_webhook.go:62]: done

File: api/v1/serviceinstance_validating_webhook.go (hunk 8 above)
Comment (TalShorSap) [reply api/v1/serviceinstance_validating_webhook.go:53]: done

File: api/v1/serviceinstance_validating_webhook.go (hunk 7 above)
Comment (TalShorSap) [reply api/v1/serviceinstance_validating_webhook.go:52]: done

File: api/v1/serviceinstance_types.go [hunk 14]
```diff
@@ -113,6 +116,9 @@ type ServiceInstanceStatus struct {
 
 	// HashedSpec is the hashed spec without the shared property
 	HashedSpec string `json:"hashedSpec,omitempty"`
+
+	// The subaccount id of the service binding
```
[the remark this answers is not in the document]
Comment (TalShorSap) [reply api/v1/serviceinstance_types.go:120]: done

File: api/v1/serviceinstance_types.go (hunk 6 above)
Comment (TalShorSap) [reply api/v1/serviceinstance_types.go:86]: done

File: sapbtp-operator-charts/templates/configmap.yml (hunk 11 above)
Comment (TalShorSap) [reply sapbtp-operator-charts/templates/configmap.yml:26]: copied from 6 lines above

File: internal/secrets/resolver.go (hunk 12 above)
Comment (TalShorSap) [reply internal/secrets/resolver.go:38]: done

File: internal/secrets/resolver.go (hunk 13 above)
Comment (TalShorSap) [reply internal/secrets/resolver.go:34]: done

**Pull request merged**